src/contracts.nit

```diff
@@ -0,0 +1,847 @@
+	var mainmodule: MModule
+
+	# Actual visited module
+	var visited_module: AModule
+
+	var ast_builder: ASTBuilder
+
+	# The `ASignature` of the actual build contract
+	var n_signature: ASignature is noinit
+
+	# The `MSignature` of the actual build contract
+	var m_signature: MSignature is noinit
+
+	# The `current_location` can corresponding of the annotation or method location.
+	var current_location: Location is noinit
+
+	# Is the contrat is an introduction or not?
+	var is_intro_contract: Bool is noinit
+
+	# Actual visited class
+	var visited_class: nullable AClassdef
+
+	# is `no_contract` annotation was found
+	var find_no_contract = false
+
+	redef fun visit(node)
+	do
+		node.create_contracts(self)
+		node.visit_all(self)
+	end
+
+	# Method use to define the signature part of the method `ASignature` and `MSignature`
+	# The given `mcontract` provided `adapt_nsignature` and `adapt_msignature` to copy and adapt the given signature (`nsignature` and `msignature`)
+	fun define_signature(mcontract: MContract, nsignature: nullable ASignature, msignature: nullable MSignature)
+	do
+		if nsignature != null and msignature != null then nsignature.ret_type = msignature.return_mtype
+		self.n_signature = mcontract.adapt_nsignature(nsignature)
+		self.m_signature = mcontract.adapt_msignature(msignature)
+	end
+
+	# Define the new contract take in consideration that an old contract exist or not
+	private fun build_contract(n_annotation: AAnnotation, mcontract: MContract, mclassdef: MClassDef): MMethodDef
+	do
+		self.current_location = n_annotation.location
+		# Retrieving the expression provided in the annotation
+		var n_condition = n_annotation.construct_condition(self)
+		var m_contractdef: AMethPropdef
+		if is_intro_contract then
+			# Create new contract method
+			m_contractdef = mcontract.create_intro_contract(self, n_condition, mclassdef)
+		else
+			# Create a redef of contract to take in consideration the new condition
+			m_contractdef = mcontract.create_subcontract(self, n_condition, mclassdef)
+		end
+		var contract_propdef = m_contractdef.mpropdef
+		# The contract has a null mpropdef, this should never happen
+		assert contract_propdef != null
+		return contract_propdef
+	end
+end
+
+# This visitor checks the `callsite` to see if the target `mpropdef` has a contract.
+private class CallSiteVisitor
+	super Visitor
+
+	var toolcontext: ToolContext
+
+	# Actual visited method
+	var visited_method: APropdef is noinit
+
+	redef fun visit(node)
+	do
+		node.check_callsite(self)
+		node.visit_all(self)
+	end
+
+	# Check if the callsite calls a method with a contract.
+	# If it's the case the callsite is replaced by another callsite to the contract method.
+	private fun drive_method_contract(callsite: CallSite): CallSite
+	do
+		if callsite.mpropdef.has_contract and not callsite.recv.need_anchor then
+			var recv = callsite.recv
+			if recv.need_anchor then recv = callsite.anchor.as(not null)
```

Comment:
not really.

anchor is the resolved context to solve a type with formal type.

eg

```nit
class Foo[E: Bar]
   fun foo do
      var x = new Array[E]
      x.add(new Bar)
      var b: Bar = x.pop # `pop` have contracts
   end
end
```

here the static type of `x` is `Array[E]` (and the static type of `self` is `Foo[E]` but it's not important here).
The problem is that `E` is a formal type, so how to reason with it, for instance: "is `var b: Bar = x.pop` valid?"
The anchor is used to reason about them. In the method, the anchor is `Foo[Bar]`, it is used to know how formal types are statically bounded.
The anchor is needed as a local context (as the mmodule) to answer some type-related questions.